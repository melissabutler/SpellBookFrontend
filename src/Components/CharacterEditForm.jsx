import React, { useContext } from "react";
import { useParams, useNavigate } from "react-router-dom";

import useFields from "../Hooks/useFields";
import CurrentCharacterContext from "../currentCharacterContext";


import { Container, Row, Col } from 'react-bootstrap';

const CharacterEditForm = ({editCharacter, toggleEdit, handleEdit}) => {
    const navigate = useNavigate();
    let character = useContext(CurrentCharacterContext)
    let { id }= useParams();

    const [formData, handleChange, resetFormData] = useFields({
        char_name: character.char_name,
        char_class: character.char_class,
        lvl: character.lvl,
        strength: character.strength,
        dexterity: character.dexterity,
        constitution: character.constitution,
        intelligence: character.intelligence,
        wisdom: character.wisdom,
        charisma: character.charisma,
})

    const handleSubmit = async e => {
        e.preventDefault();
  
        let updatedCharacter = {
            "char_name": formData.char_name.toString(),
            "char_class": formData.char_class.toString(),
            "lvl": parseInt(formData.lvl),
            "strength": parseInt(formData.strength),
            "dexterity": parseInt(formData.dexterity),
            "constitution": parseInt(formData.constitution),
            "intelligence": parseInt(formData.intelligence),
            "wisdom": parseInt(formData.wisdom),
            "charisma": parseInt(formData.charisma)
        }
        
        await editCharacter({updatedCharacter}, id)
        toggleEdit();
    }


     /** This piece condenses the options list for numeric stats, in this case character level and ability scores.
     * As they both are a range from 1-20, they can share this list. 
    */
     let statRange = [];

     for(let i = 0; i <= 20; i++){
         let key = "level-" + i;
         statRange.push(<option key={key} value={i} >{i}</option>)
     }

    return (
        <Container className="CharacterForm">
            <Row>
                <h1>Edit Character</h1>
            </Row>
             <form className="CharacterForm-form" >
            <Row>
                    <Col className="CharacterForm-section">
                        <h5 htmlFor="char_name" className="CharacterForm-h4">Character Name</h5>
                    
                        <input
                            className="CharacterForm-input"
                            id="char_name"
                            name="char_name"
                            type="text"
                            value={formData.char_name}
                            onChange={handleChange}
                            placeholder="Character Name"
                        />
                    </Col>
                    
                    <Col className="CharacterForm-section">
                        <h5 className="CharacterForm-h4" htmlFor="char_class">Character Class</h5>
                   
                    <select id="char_class" value={formData.char_class} name="char_class" onChange={handleChange}>
                        <option key="placeholder" value="">Select a class</option>
                        <option key="bard" value="bard">Bard</option>
                        <option key="cleric" value="cleric">Cleric</option>
                        <option key="druid" value="druid">Druid</option>
                        <option key="paladin" value="paladin">Paladin</option>
                        <option key="ranger" value="ranger">Ranger</option>
                        <option key="sorcerer" value="sorcerer">Sorcerer</option>
                        <option key="warlock" value="warlock">Warlock</option>
                        <option key="wizard" value="wizard">Wizard</option>
                    </select>
                    </Col>
                    <Col className="CharacterForm-section">
                        <h5 className="CharacterForm-h4" htmlFor="lvl">Character Level</h5>
                        
                        <select id="lvl" value={formData.lvl} name="lvl" onChange={handleChange}>
                            <option key="placeholder" value="">Level</option>
                            {statRange}
                        </select>
                    </Col>
            </Row>

            <Row className="CharacterForm-section-ability">
                <Col className="CharacterForm-section">
                        <h4 className="CharacterForm-h4" htmlFor="strength">Strength</h4>
                <select className="ability-score" value={formData.strength} name="strength" onChange={handleChange}>
                    <option key="placeholder" value="">STR</option>
                    {statRange}
                </select>
                </Col>

                <Col className="CharacterForm-section">
                        <h4 className="CharacterForm-h4" htmlFor="dexterity">Dexterity</h4>
                <select className="ability-score" value={formData.dexterity} name="dexterity" onChange={handleChange}>
                    <option key="placeholder" value="">DEX</option>
                    {statRange}
                </select>
                </Col>

                <Col className="CharacterForm-section">
                        <h4 className="CharacterForm-h4" htmlFor="constitution">Constitution</h4>
                <select className="ability-score" value={formData.constitution} name="constitution" onChange={handleChange}>
                    <option key="placeholder" value="">CON</option>
                    {statRange}
                </select>
                </Col>

                <Col className="CharacterForm-section">
                        <h4 className="CharacterForm-h4" htmlFor="intelligence">Intelligence</h4>
                <select className="ability-score" value={formData.intelligence} name="intelligence" onChange={handleChange}>
                    <option key="placeholder" value="">INT</option>
                    {statRange}
                </select>
                </Col>

                <Col className="CharacterForm-section">
                        <h4 className="CharacterForm-h4" htmlFor="wisdom">Wisdom</h4>
                <select className="ability-score" value={formData.wisdom} name="wisdom" onChange={handleChange}>
                    <option key="placeholder" value="">WIS</option>
                    {statRange}
                </select>
                </Col>

                <Col className="CharacterForm-section">
                        <h4 className="CharacterForm-h4" htmlFor="charisma">Charisma</h4>
                <select className="ability-score" value={formData.charisma} name="charisma" onChange={handleChange}>
                    <option key="placeholder" value="">CHA</option>
                    {statRange}
                </select>
                </Col>
           
            </Row>
            <button onClick={handleSubmit}>Submit</button>
            <button onClick={handleEdit}>Hide Edit</button>
            </form>
            
        </Container>
    )
    

}

export default CharacterEditForm;