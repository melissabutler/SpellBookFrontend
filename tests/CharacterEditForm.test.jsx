import React from "react";
import { describe, it, expect, vi } from "vitest";
import { fireEvent, render, screen} from '@testing-library/react'
import CharacterEditForm from "../src/Components/CharacterEditForm";

describe("CharacterEditForm", function() {
    vi.mock('react', async (importOriginal) => {
        const react = await importOriginal('react');
        return {
            ...react,
            useContext: () => ({ currentUser: {
                username: 'u1',
                email: 'email@email.com',
                isAdmin: false,
                characters: [
                    {
                    char_class: "cleric",
                    char_name: "Beebo",
                    id: 1,
                    lvl: 3
                },
                {
                    char_class: "cleric",
                    char_name: "Beebo",
                    id: 2,
                    lvl: 3
                }
                ]
            }})
    }})

    vi.mock('react-router-dom', async () => {
        const mod = await vi.importActual('react-router-dom');
        return {
            ...mod, 
            useParams: () => ({
                id: 1
            }),
            useNavigate: () => "/"
        }
    })


    it('renders the CharacterEditForm component', async () => {
        render(<CharacterEditForm />)
        expect(screen.getByText('Edit Character')).toBeInTheDocument();
      
    });

    it('matches snapshot', () => {
        const { asFragment } = render(<CharacterEditForm />);
        expect(asFragment()).toMatchSnapshot();
    })


})
   