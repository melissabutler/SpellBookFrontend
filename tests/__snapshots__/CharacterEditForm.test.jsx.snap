// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CharacterEditForm > matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="CharacterForm container"
  >
    <div
      class="row"
    >
      <h1>
        Edit Character
      </h1>
    </div>
    <form
      class="CharacterForm-form"
    >
      <div
        class="row"
      >
        <div
          class="CharacterForm-section col"
        >
          <h5
            class="CharacterForm-h4"
            for="char_name"
          >
            Character Name
          </h5>
          <input
            class="CharacterForm-input"
            id="char_name"
            name="char_name"
            placeholder="Character Name"
            type="text"
            value=""
          />
        </div>
        <div
          class="CharacterForm-section col"
        >
          <h5
            class="CharacterForm-h4"
            for="char_class"
          >
            Character Class
          </h5>
          <select
            id="char_class"
            name="char_class"
          >
            <option
              value=""
            >
              Select a class
            </option>
            <option
              value="bard"
            >
              Bard
            </option>
            <option
              value="cleric"
            >
              Cleric
            </option>
            <option
              value="druid"
            >
              Druid
            </option>
            <option
              value="paladin"
            >
              Paladin
            </option>
            <option
              value="ranger"
            >
              Ranger
            </option>
            <option
              value="sorcerer"
            >
              Sorcerer
            </option>
            <option
              value="warlock"
            >
              Warlock
            </option>
            <option
              value="wizard"
            >
              Wizard
            </option>
          </select>
        </div>
        <div
          class="CharacterForm-section col"
        >
          <h5
            class="CharacterForm-h4"
            for="lvl"
          >
            Character Level
          </h5>
          <select
            id="lvl"
            name="lvl"
          >
            <option
              value=""
            >
              Level
            </option>
            <option
              value="0"
            >
              0
            </option>
            <option
              value="1"
            >
              1
            </option>
            <option
              value="2"
            >
              2
            </option>
            <option
              value="3"
            >
              3
            </option>
            <option
              value="4"
            >
              4
            </option>
            <option
              value="5"
            >
              5
            </option>
            <option
              value="6"
            >
              6
            </option>
            <option
              value="7"
            >
              7
            </option>
            <option
              value="8"
            >
              8
            </option>
            <option
              value="9"
            >
              9
            </option>
            <option
              value="10"
            >
              10
            </option>
            <option
              value="11"
            >
              11
            </option>
            <option
              value="12"
            >
              12
            </option>
            <option
              value="13"
            >
              13
            </option>
            <option
              value="14"
            >
              14
            </option>
            <option
              value="15"
            >
              15
            </option>
            <option
              value="16"
            >
              16
            </option>
            <option
              value="17"
            >
              17
            </option>
            <option
              value="18"
            >
              18
            </option>
            <option
              value="19"
            >
              19
            </option>
            <option
              value="20"
            >
              20
            </option>
          </select>
        </div>
      </div>
      <div
        class="CharacterForm-section-ability row"
      >
        <div
          class="CharacterForm-section col"
        >
          <h4
            class="CharacterForm-h4"
            for="strength"
          >
            Strength
          </h4>
          <select
            class="ability-score"
            name="strength"
          >
            <option
              value=""
            >
              STR
            </option>
            <option
              value="0"
            >
              0
            </option>
            <option
              value="1"
            >
              1
            </option>
            <option
              value="2"
            >
              2
            </option>
            <option
              value="3"
            >
              3
            </option>
            <option
              value="4"
            >
              4
            </option>
            <option
              value="5"
            >
              5
            </option>
            <option
              value="6"
            >
              6
            </option>
            <option
              value="7"
            >
              7
            </option>
            <option
              value="8"
            >
              8
            </option>
            <option
              value="9"
            >
              9
            </option>
            <option
              value="10"
            >
              10
            </option>
            <option
              value="11"
            >
              11
            </option>
            <option
              value="12"
            >
              12
            </option>
            <option
              value="13"
            >
              13
            </option>
            <option
              value="14"
            >
              14
            </option>
            <option
              value="15"
            >
              15
            </option>
            <option
              value="16"
            >
              16
            </option>
            <option
              value="17"
            >
              17
            </option>
            <option
              value="18"
            >
              18
            </option>
            <option
              value="19"
            >
              19
            </option>
            <option
              value="20"
            >
              20
            </option>
          </select>
        </div>
        <div
          class="CharacterForm-section col"
        >
          <h4
            class="CharacterForm-h4"
            for="dexterity"
          >
            Dexterity
          </h4>
          <select
            class="ability-score"
            name="dexterity"
          >
            <option
              value=""
            >
              DEX
            </option>
            <option
              value="0"
            >
              0
            </option>
            <option
              value="1"
            >
              1
            </option>
            <option
              value="2"
            >
              2
            </option>
            <option
              value="3"
            >
              3
            </option>
            <option
              value="4"
            >
              4
            </option>
            <option
              value="5"
            >
              5
            </option>
            <option
              value="6"
            >
              6
            </option>
            <option
              value="7"
            >
              7
            </option>
            <option
              value="8"
            >
              8
            </option>
            <option
              value="9"
            >
              9
            </option>
            <option
              value="10"
            >
              10
            </option>
            <option
              value="11"
            >
              11
            </option>
            <option
              value="12"
            >
              12
            </option>
            <option
              value="13"
            >
              13
            </option>
            <option
              value="14"
            >
              14
            </option>
            <option
              value="15"
            >
              15
            </option>
            <option
              value="16"
            >
              16
            </option>
            <option
              value="17"
            >
              17
            </option>
            <option
              value="18"
            >
              18
            </option>
            <option
              value="19"
            >
              19
            </option>
            <option
              value="20"
            >
              20
            </option>
          </select>
        </div>
        <div
          class="CharacterForm-section col"
        >
          <h4
            class="CharacterForm-h4"
            for="constitution"
          >
            Constitution
          </h4>
          <select
            class="ability-score"
            name="constitution"
          >
            <option
              value=""
            >
              CON
            </option>
            <option
              value="0"
            >
              0
            </option>
            <option
              value="1"
            >
              1
            </option>
            <option
              value="2"
            >
              2
            </option>
            <option
              value="3"
            >
              3
            </option>
            <option
              value="4"
            >
              4
            </option>
            <option
              value="5"
            >
              5
            </option>
            <option
              value="6"
            >
              6
            </option>
            <option
              value="7"
            >
              7
            </option>
            <option
              value="8"
            >
              8
            </option>
            <option
              value="9"
            >
              9
            </option>
            <option
              value="10"
            >
              10
            </option>
            <option
              value="11"
            >
              11
            </option>
            <option
              value="12"
            >
              12
            </option>
            <option
              value="13"
            >
              13
            </option>
            <option
              value="14"
            >
              14
            </option>
            <option
              value="15"
            >
              15
            </option>
            <option
              value="16"
            >
              16
            </option>
            <option
              value="17"
            >
              17
            </option>
            <option
              value="18"
            >
              18
            </option>
            <option
              value="19"
            >
              19
            </option>
            <option
              value="20"
            >
              20
            </option>
          </select>
        </div>
        <div
          class="CharacterForm-section col"
        >
          <h4
            class="CharacterForm-h4"
            for="intelligence"
          >
            Intelligence
          </h4>
          <select
            class="ability-score"
            name="intelligence"
          >
            <option
              value=""
            >
              INT
            </option>
            <option
              value="0"
            >
              0
            </option>
            <option
              value="1"
            >
              1
            </option>
            <option
              value="2"
            >
              2
            </option>
            <option
              value="3"
            >
              3
            </option>
            <option
              value="4"
            >
              4
            </option>
            <option
              value="5"
            >
              5
            </option>
            <option
              value="6"
            >
              6
            </option>
            <option
              value="7"
            >
              7
            </option>
            <option
              value="8"
            >
              8
            </option>
            <option
              value="9"
            >
              9
            </option>
            <option
              value="10"
            >
              10
            </option>
            <option
              value="11"
            >
              11
            </option>
            <option
              value="12"
            >
              12
            </option>
            <option
              value="13"
            >
              13
            </option>
            <option
              value="14"
            >
              14
            </option>
            <option
              value="15"
            >
              15
            </option>
            <option
              value="16"
            >
              16
            </option>
            <option
              value="17"
            >
              17
            </option>
            <option
              value="18"
            >
              18
            </option>
            <option
              value="19"
            >
              19
            </option>
            <option
              value="20"
            >
              20
            </option>
          </select>
        </div>
        <div
          class="CharacterForm-section col"
        >
          <h4
            class="CharacterForm-h4"
            for="wisdom"
          >
            Wisdom
          </h4>
          <select
            class="ability-score"
            name="wisdom"
          >
            <option
              value=""
            >
              WIS
            </option>
            <option
              value="0"
            >
              0
            </option>
            <option
              value="1"
            >
              1
            </option>
            <option
              value="2"
            >
              2
            </option>
            <option
              value="3"
            >
              3
            </option>
            <option
              value="4"
            >
              4
            </option>
            <option
              value="5"
            >
              5
            </option>
            <option
              value="6"
            >
              6
            </option>
            <option
              value="7"
            >
              7
            </option>
            <option
              value="8"
            >
              8
            </option>
            <option
              value="9"
            >
              9
            </option>
            <option
              value="10"
            >
              10
            </option>
            <option
              value="11"
            >
              11
            </option>
            <option
              value="12"
            >
              12
            </option>
            <option
              value="13"
            >
              13
            </option>
            <option
              value="14"
            >
              14
            </option>
            <option
              value="15"
            >
              15
            </option>
            <option
              value="16"
            >
              16
            </option>
            <option
              value="17"
            >
              17
            </option>
            <option
              value="18"
            >
              18
            </option>
            <option
              value="19"
            >
              19
            </option>
            <option
              value="20"
            >
              20
            </option>
          </select>
        </div>
        <div
          class="CharacterForm-section col"
        >
          <h4
            class="CharacterForm-h4"
            for="charisma"
          >
            Charisma
          </h4>
          <select
            class="ability-score"
            name="charisma"
          >
            <option
              value=""
            >
              CHA
            </option>
            <option
              value="0"
            >
              0
            </option>
            <option
              value="1"
            >
              1
            </option>
            <option
              value="2"
            >
              2
            </option>
            <option
              value="3"
            >
              3
            </option>
            <option
              value="4"
            >
              4
            </option>
            <option
              value="5"
            >
              5
            </option>
            <option
              value="6"
            >
              6
            </option>
            <option
              value="7"
            >
              7
            </option>
            <option
              value="8"
            >
              8
            </option>
            <option
              value="9"
            >
              9
            </option>
            <option
              value="10"
            >
              10
            </option>
            <option
              value="11"
            >
              11
            </option>
            <option
              value="12"
            >
              12
            </option>
            <option
              value="13"
            >
              13
            </option>
            <option
              value="14"
            >
              14
            </option>
            <option
              value="15"
            >
              15
            </option>
            <option
              value="16"
            >
              16
            </option>
            <option
              value="17"
            >
              17
            </option>
            <option
              value="18"
            >
              18
            </option>
            <option
              value="19"
            >
              19
            </option>
            <option
              value="20"
            >
              20
            </option>
          </select>
        </div>
      </div>
      <button>
        Submit
      </button>
      <button>
        Hide Edit
      </button>
    </form>
  </div>
</DocumentFragment>
`;
